### PR TITLE
Update MoCo tutorial to use low-level blocks

### PR DESCRIPTION
**Pull request by guarin**
- Rewrite MocoModel
- Remove `MocoModel.forward` as unused
- Switch to terminology from paper
- Rewrite Classifier
- Remove normalization in `Classifier.forward` as it is not in protocol
- Remove `pl.metrics.Accuracy` as it is deprecated
- Add code for accuracy calculation

I decided to use the non-symmetric version of MoCo due to #564 and because it is easier to understand.

[moco_tutorial.pdf](https://github.com/lightly-ai/lightly/files/7564322/moco_tutorial.pdf)

### CODE REVIEW

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py [hunk 1]
```diff
@@ -208,52 +215,72 @@
...
+        q = self.backbone(x_q).flatten(start_dim=1)
+        q = self.projection_head(q)
+
+        # get keys
+        k, shuffle = batch_shuffle(x_k)
+        k = self.backbone_momentum(k).flatten(start_dim=1)
+        k = self.projection_head_momentum(k)
+        k = batch_unshuffle(k, shuffle)
+
+        loss = self.criterion(q, k)
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_moco_memory_bank.py:260]: Do I assume correctly, that this q, k, are the same ones as used in this image?
https://user-images.githubusercontent.com/11435359/71603927-0ca98d00-2b14-11ea-9fd8-10d984a2de45.png

If yes, could you add this image to the tutorial? I think it would make it easier to understand these steps.

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py [hunk 2]
```diff
@@ -208,52 +215,72 @@
...
 
-        # create a moco based on ResNet
-        self.resnet_moco = \
-            lightly.models.MoCo(backbone, num_ftrs=512, m=0.99, batch_shuffle=True)
+        # create a moco model based on ResNet
+        self.projection_head = MoCoProjectionHead(512, 512, 128)
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+        deactivate_requires_grad(self.backbone_momentum)
+        deactivate_requires_grad(self.projection_head_momentum)
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_moco_memory_bank.py:234]: Nice! This shows that MoCo really has those 4 building blocks and how they belong together.

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py [hunk 3]
```diff
@@ -208,52 +215,72 @@
...
-        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optim, max_epochs)
+        optim = torch.optim.SGD(
+            self.parameters(),
+            lr=6e-2,
+            momentum=0.9,
+            weight_decay=5e-4,
+        )
+        scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optim, max_epochs
+        )
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_moco_memory_bank.py:283]: Well formatted! I love it.

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py [hunk 4]
```diff
@@ -264,58 +291,69 @@ def configure_optimizers(self):
...
+
+    def validation_epoch_end(self, outputs):
+        # calculate and log top1 accuracy
+        if outputs:
+            total_num = torch.Tensor([0],).to(self.device)
+            total_correct = torch.Tensor([0.0],).to(self.device)
+            for num, correct in outputs:
+                total_num += num
+                total_correct += correct
+            acc = total_correct.item() / total_num.item()
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_moco_memory_bank.py:349]: What is the reason here to use a GPU tensor for two single scalars?

Wouldn't it be easier to compute it in plain python?
```python
total_num =0
total_correct = 0
for num, correct in outputs:
   total_num += num.item()
   total_correct += correct.item()
acc = total_correct / total_item
```

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py (hunk 1 above)
Comment (guarin) [reply docs/source/tutorials_source/package/tutorial_moco_memory_bank.py:260]: Maybe we should discuss whether we want to add these images to the tutorials in general (SimCLR, SimSiam, ...). I feel that the tutorials are already quite long and would let people have a look at the paper if they are interested.

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py (hunk 4 above)
Comment (guarin) [reply docs/source/tutorials_source/package/tutorial_moco_memory_bank.py]: Yes you are correct! I updated the pr accordingly.

File: docs/source/tutorials_source/package/tutorial_moco_memory_bank.py (hunk 1 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/package/tutorial_moco_memory_bank.py:260]: Ok, then let's leave them away for the moment.

**Comment by philippmwirth** [comment]
This looks great! Does it work? What's the accuracy you get when you train longer?

**Review by MalteEbner — APPROVED** [review]
From my side, this is ready to be merged concerning the code. However, ensure that it works before merging it.

**Comment by guarin** [comment]
> This looks great! Does it work? What's the accuracy you get when you train longer?

I got 75.4% compared to 70.7% in the current version.

**Review by philippmwirth — APPROVED** [review]
